databricks/koalas/frame.py

```diff
@@ -1111,6 +1112,35 @@ def to_koalas(self):
         else:
             return DataFrame(self)
 
+    @contextmanager
+    def cache(self):
+        """
+        Yields and caches the current DataFrame as a Spark DataFrame.
```

Comment:
I think this should return a Koalas DataFrame with keeping metadata as is

Reply:
@HyukjinKwon is this implementation good enough? I think you were involved in that comment for pyspark: 
https://github.com/apache/spark/pull/14579#discussion_r74813935

Personally, I think the situation that was discussed in the PR is an optimization, but you have more context than I do.

Reply:
ah, I think that's fine for now. Users are already able to call `cache` multiple times. It will show an error that:

```
19/05/29 13:25:43 WARN CacheManager: Asked to cache already cached data.
```

Reply:
@HyukjinKwon, @thunterdb, do you guys want me to implement it like this -

```
def __enter__(self):
        return self

def __exit__(self):
        self.unpersist()
```

Or the current implementation is fine?

Reply:
> I think this should return a Koalas DataFrame with keeping metadata as is

@HyukjinKwon, I have a follow-up doubt. In this cache function - 
- I have to convert the koalas DataFrame into Spark DataFrame.
- Then I have to cache that Spark DataFrame and converting the cached DataFrame to koalas dataframe.
- Finally, I have to uncache the Spark DataFrame.

Am I correct or there is some mistake at my end?

Reply:
If we call `cache()`, it should return Koalas dataframe because we're calling it against Koalas DataFrame.

I was roughly thinking like:

```
koalas_df = ...
koalas_df._sdf = koalas_df._sdf.cache()
return self
```

Reply:
I was thinking like:

```python
        try:
            self._sdf = self._sdf.cache()
            yield self
        finally:
            self._sdf.unpersist()
```

Does this work?

Also, can we make this `cache` API acts like for both normal API call:

```python
df = df.cache()
```

and also context manager:

```
with df.cache() as df:
    ....
```

?

Reply:
In my current implementation, I am only able to use it with `with` syntax.

Reply:
@HyukjinKwon, The `try-finally` block won't let us use this
```
df = df.cache()
```
Since it only `yields` the df but not `returns` it.
I have changed my implementation. Please have a look.